Add ISO central and peripheral security handling

Thalley opened the pull request:
The `required_sec_level` from `struct bt_iso_chan` and the `sec_level` from `struct bt_iso_server` was added to the API long time ago, but was never implemented properly. This PR updates the ISO implementation to properly use these values. 

fixes #33865

Thalley committed:
Bluetooth: ISO: Add server security check

Add check for connection and server security
levels, and reject the CIS if the ACL security
level is too low.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Central security request

If the required_sec_level is lower than the
conn->sec_level, the central will now initialize the
security procecure to ensure that the CIS is encrypted
properly.

The algorithm implemented is as follows:
1) Check security levels for each (acl, iso) pair
2) For those with insufficient security,
   call bt_conn_set_security
3) For those with sufficient security, connect the CIS
4) Once the ISO from 2) has been encrypted, connect the
   CIS for the specific ACL

The idea behind this was to implement similar support
for autonomous encryption as we have for L2CAP.
It is more complex for ISO as we are dealing with
an array of (acl, iso) pairs, meaning more can go
wrong.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Casper-Bonde-Bose reviewed (dismissed):
LGTM

Thalley committed:
Bluetooth: ISO: Guard sec_level with CONFIG_BT_SMP

The ISO security implementation works by verifying
against the acl (bt_conn) sec_level field. The
bt_conn sec_level field is only available
if CONFIG_BT_SMP is enabled, so this commit
adds guards for all ISO security checks as well.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

alexsven commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1441,6 +1480,92 @@ int bt_iso_cig_terminate(struct bt_iso_cig *cig)
+	}
+
+	if (param_count == 0) {
+		/* Nothing to do; this shouldn't happen */
```
If this shouldn't happen, should there be a BT_WRN or BT_ERR here?

Thalley replied:
Added a `BT_ERR`

Thalley replied:
Remove the `BT_ERR` again, and update the comment. 

This can actually happen if no CIS are pending encryption, but the ACL is encrypted by other means.

alexsven commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1441,6 +1480,92 @@ int bt_iso_cig_terminate(struct bt_iso_cig *cig)
+		bt_conn_set_state(iso_chan->iso, BT_CONN_CONNECT);
+		bt_iso_chan_set_state(iso_chan, BT_ISO_CONNECT);
+
+		__ASSERT(cig != NULL, "CIG was NULL");
```
Not super important since it is an ASSERT, but would make more sense to have it before setting the states

alexsven commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1458,21 +1583,88 @@ static int hci_le_create_cis(const struct bt_iso_connect_param *param,
 	memset(req, 0, sizeof(*req));
 
-	req->num_cis = count;
+	req->num_cis = 0;
```
Is this necessary when req is memset to 0 two lines above?

alexsven commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1458,21 +1583,88 @@ static int hci_le_create_cis(const struct bt_iso_connect_param *param,
+	}
+
+	/* If all CIS are pending for security, do nothing,
+	 * but return an recognizable return value
```
'a recognizable return value'

alexsven commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1458,21 +1583,88 @@ static int hci_le_create_cis(const struct bt_iso_connect_param *param,
+static int iso_chan_connect_security(const struct bt_iso_connect_param *param,
+				     size_t count)
+{
+	/* conn_idx_handled is a an array of booleans for which conn indexes
```
a an array...

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -442,15 +444,16 @@ static void bt_iso_chan_disconnected(struct bt_iso_chan *chan, uint8_t reason)
-			SYS_SLIST_FOR_EACH_CONTAINER(&cig->cis_channels, chan, node) {
-				if (chan->state == BT_ISO_CONNECTED ||
-				chan->state == BT_ISO_CONNECT) {
+			SYS_SLIST_FOR_EACH_CONTAINER(&cig->cis_channels, tmp, node) {
```
`SYS_SLIST_FOR_EACH_CONTAINER` will assign each item in the list to `tmp`. If you are familiar with something like python, this would be similar to `for tmp in cig->cis_channels`

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1441,6 +1480,92 @@ int bt_iso_cig_terminate(struct bt_iso_cig *cig)
+	}
+
+	param_count = 0;
+	for (size_t i = 0; i < ARRAY_SIZE(iso_conns); i++) {
```
`ARRAY_SIZE` returns a `size_t` value, so it is to compare values of the same type.

Thalley commented:
Seeing some odd behavior in the shell, so marking as DNM.

Thalley committed:
Bluetooth: Shell: Add security level option to "iso connect"

The "iso connect" now also takes a security level option like
"iso listen", so that the central can easily encrypt
and connect a CIS in a single command.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -116,9 +118,12 @@ static int cmd_listen(const struct shell *sh, size_t argc, char *argv[])
 		return -ENOEXEC;
 	}
 
+
```
minor: unnecessary empty line here

Thalley replied:
Fixed

jhedberg reviewed (dismissed):
Looks generally ok to me. One tiny whitespace issue pointed out as a separate comment, which would be good to fix (I'm approving anyway).

rick1082 commented on samples/bluetooth/peripheral_iso/src/main.c:
```diff
@@ -141,7 +141,9 @@ static int iso_accept(const struct bt_iso_accept_info *info,
 
 static struct bt_iso_server iso_server = {
+#if defined(CONFIG_BT_SMP)
 	.sec_level = BT_SECURITY_L1,
```
Should we use BT_SECURITY_L2 once if enabled CONFIG_BT_SMP?

Thalley committed:
Bluetooth: ISO: Move iso_server NULL check for inc. reqs

Move the NULL-check for the iso_server to an earlier point.
There's no need to do anything on the device if no iso_server
has been registered.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1512,6 +1518,94 @@ int bt_iso_cig_terminate(struct bt_iso_cig *cig)
+		bt_iso_chan_set_state(iso_chan, BT_ISO_DISCONNECTED);
+
+		if (hci_status == BT_HCI_ERR_SUCCESS) {
+			param[param_count].acl = acl;
```
Great observation!

`param` should have used `CONFIG_BT_ISO_MAX_CHAN`. Updated now.

asbjornsabo commented on subsys/bluetooth/host/iso.c:
```diff
@@ -959,12 +961,26 @@ int hci_le_accept_cis(uint16_t handle)
+		return BT_HCI_ERR_SUCCESS;
+	}
+
+	if (conn->sec_level >= iso_server->sec_level) {
```
I believe this is now fixed in the last commit, which moves the iso_server check to before the call to this function.

asbjornsabo reviewed (dismissed):
My comments have been addressed.

Thalley commented:
Rebased to solve merge conflict

Thalley commented:
Rebased to solve merge conflict

Thalley commented:
Rebased to solve merge conflict

Casper-Bonde-Bose reviewed (approved):
LGTM
